clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -0,0 +1,191 @@
+                    final String mech = "GSSAPI";
+
+                    LOG.debug("serviceHostname is '" + serviceHostname + "'");
+                    LOG.debug("servicePrincipalName is '" + servicePrincipalName + "'");
+                    LOG.debug("SASL mechanism(mech) is '" + mech + "'");
+                    boolean usingNativeJgss = Boolean.getBoolean("sun.security.jgss.native");
+                    if (usingNativeJgss) {
+                        try {
+                            GSSManager manager = GSSManager.getInstance();
+                            Oid krb5Mechanism = new Oid("1.2.840.113554.1.2.2");
+                            GSSName gssName = manager.createName(servicePrincipalName + "@" + serviceHostname, GSSName.NT_HOSTBASED_SERVICE);
+                            GSSCredential cred = manager.createCredential(gssName, GSSContext.INDEFINITE_LIFETIME, krb5Mechanism, GSSCredential.ACCEPT_ONLY);
+                            subject.getPrivateCredentials().add(cred);
+                        } catch (GSSException ex) {
+                            LOG.warn("Cannot add private credential to subject; clients authentication may fail", ex);
+                        }
+                    }
+
+                    try {
+                        return Subject.doAs(subject, new PrivilegedExceptionAction<SaslServer>() {
+                                public SaslServer run() {
+                                    try {
+                                        SaslServer saslServer;
+                                        saslServer = Sasl.createSaslServer(mech, servicePrincipalName, serviceHostname, null, saslServerCallbackHandler);
+                                        return saslServer;
+                                    } catch (SaslException e) {
+                                        LOG.error("Kafka Server failed to create a SaslServer to interact with a client during session authentication: " + e);
+                                        return null;
+                                    }
+                                }
+                            });
+                    } catch (PrivilegedActionException e) {
+                        LOG.error("KafkaBroker experienced a PrivilegedActionException exception while creating a SaslServer using a JAAS principal context:" + e);
+                    }
+                } catch (IndexOutOfBoundsException e) {
+                    LOG.error("Kafka Server principal name/hostname determination error: ", e);
+                }
+            }
+        }
+        return null;
+    }
+
+    public void authenticate() throws IOException {
+        if (netOutBuffer != null && !flushNetOutBuffer()) {
+            transportLayer.addInterestOps(SelectionKey.OP_WRITE);
+            return;
+        }
+
+        if (saslServer.isComplete()) {
+            transportLayer.addInterestOps(SelectionKey.OP_READ);
```

Comment:
Harsha changed this to be:

```transportLayer.removeInterestOps(SelectionKey.OP_WRITE);```

I think this also addresses the `turn off OP_WRITE` comment.